edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/ServiceIdentityTree.cs

```diff
@@ -0,0 +1,275 @@
+                newNode.Parent = parentNode;
+                parentNode.ForEach(node => node.Children.Add(newNode));
+            }
+
+            // Check if there are any dangling children that can now be hooked up,
+            // this include placing the new node as the new root.
+            List<ServiceIdentityTreeNode> danglingChildren =
+                nodes
+                .Select(kvp => kvp.Value)
+                .Where(s => s.Identity.ParentScopes.Count() > 0 && s.Identity.ParentScopes.Contains(device.DeviceScope.OrDefault()))
+                .ToList();
+
+            foreach (ServiceIdentityTreeNode child in danglingChildren)
+            {
+                child.Parent = Option.Some(newNode);
+                newNode.Children.Add(child);
+            }
+        }
+
+        void RemoveSingleNode_NoCallback(IDictionary<string, ServiceIdentityTreeNode> nodes, string id)
+        {
+            if (nodes.TryGetValue(id, out ServiceIdentityTreeNode target))
+            {
+                // Unhook from the parent
+                target.Parent.ForEach(parent => parent.Children.Remove(target));
+
+                // Unhook the children
+                target.Children.ForEach(child => child.Parent = Option.None<ServiceIdentityTreeNode>());
+
+                // Remove
+                nodes.Remove(id);
+            }
+        }
+
+        void RemoveRecursive(IDictionary<string, ServiceIdentityTreeNode> nodes, string id)
+        {
+            if (nodes.TryGetValue(id, out ServiceIdentityTreeNode target))
+            {
+                // Unhook from the parent
+                target.Parent.ForEach(parent => parent.Children.Remove(target));
+
+                // Recursively remove each child
+                var childrenSnapshot = new List<ServiceIdentityTreeNode>(target.Children);
+                childrenSnapshot.ForEach(child => this.RemoveRecursive(nodes, child.Identity.Id));
+
+                // Remove self after recursion
+                nodes.Remove(id);
+
+                // Notify callbacks
+                this.ServiceIdentityRemoved?.Invoke(this, id);
+            }
+        }
+
+        Option<ServiceIdentityTreeNode> FindDeviceByScopeId(IDictionary<string, ServiceIdentityTreeNode> nodes, string scopeId)
+        {
+            Preconditions.CheckNonWhiteSpace(scopeId, nameof(scopeId));
+
+            // Look for Edge devices with a matching scope ID
+            List<ServiceIdentityTreeNode> devices =
+                nodes
+                .Select(kvp => kvp.Value)
+                .Where(serviceIdentity => serviceIdentity.Identity.IsEdgeDevice && serviceIdentity.Identity.DeviceScope.Contains(scopeId))
+                .ToList();
+
+            if (devices.Count() > 0)
+            {
+                // There shouldn't be more than one device, but even if there is
+                // for some reason, we can just use the first one
+                return Option.Some(devices.First());
```

Comment:
I think it is better to have a data structure to protect adding a service identity with duplicate device scope.

Reply:
yea we're already protected by using a dictionary